src/terms/nonlocal.jl

```diff
@@ -178,22 +178,34 @@ end
 """
 Build form factors (Fourier transforms of projectors) for an atom centered at 0.
 """
-function build_form_factors(psp, qs)
-    qnorms = norm.(qs)
-    T = real(eltype(qnorms))
-    # Compute position-independent form factors
-    form_factors = zeros(Complex{T}, length(qs), count_n_proj(psp))
-    count = 1
-    for l in 0:psp.lmax, m in -l:l
-        prefac_lm = im^l .* ylm_real.(l, m, qs)
-        n_proj_l = size(psp.h[l + 1], 1)
+function build_form_factors(psp, G_plus_ks)
+    T = real(eltype(first(G_plus_ks)))
+
+    n_proj_max = maximum(l -> count_n_proj_radial(psp, l), 0:psp.lmax; init=0)
+    radials = IdDict{T,Matrix{T}}()
+    for Gpk in G_plus_ks
+        q = norm(Gpk)
+        if !haskey(radials, q)
+            radials_q = Matrix{T}(undef, n_proj_max, psp.lmax + 1)
+            for l in 0:psp.lmax, iproj_l in 1:count_n_proj_radial(psp, l)
+                radials_q[iproj_l, l+1] = eval_psp_projector_fourier(psp, iproj_l, l, q)
+            end
+            radials[q] = radials_q
+        end
+    end
```

Comment:
btw this is all by-hand memoization. Do we want to just slap a `@memoize` in front of `eval_psp_projector_fourier(psp, ::Real)`? What I don't really like about it is that 1 it's tricky with threading 2 it's tricky to get the memory back. The manual solution is probably simpler...

Reply:
BTW you can get this to thread using ThreadSafeDict

Reply:
> @memoize

We used that at some point and it caused trouble because of hiding some details exactly as you say @antoine-levitt . I vote against it.